### DISCUSSION TITLE
Authorization и Cookie

### DISCUSSION THREAD

**Discussion by vasyan35** [lowfc/boosty_downloader#12] [Q&A] [answered]
Здравствуйте! В конфигурации застрял на пункте post, не могу найти его. Возможно разработчики изменили код сайта. Где найти Authorization и Cookie, подскажите пожалуйста.

_#Пример для chrome:
    Перейдите на вашу страницу на boosty.to
    Нажимте клавишу F12
    Обновите страницу с помощью Shift + F5
    Перейдите в раздел Network
    Прокрутите страницу вниз
    **В списке запросов, выберите пункт post/**
    В блоке Request Headers найдите Authorization и Cookie, а затем скопируйте их в соответствующие поля в config.yml#_

**Comment by lowfc** [comment, marked as answer]
Здравствуйте!
1. Перейдите на страницу boosty.to;
2. Войдите в свой аккаунт на boosty.to;
3. Нажмите клавишу F12;
4. Перейдите в раздел Console (Консоль);
4. В поле ввода консоли введите следующий код и нажмите Enter:

```
const full_cookie=`${document.cookie}`;function getCookie(o,e){const l=e.split(`; ${o}=`);if(2===l.length)return l.pop().split(";").shift()}function parseAuthCookie(o){if(void 0===o)return;let e=o.replaceAll("%7B","{");e=e.replaceAll("%22",'"'),e=e.replaceAll("%3A",":"),e=e.replaceAll("%2C",","),e=e.replaceAll("%7D","}");let l=JSON.parse(e).accessToken;return void 0!==l?"Bearer "+l:void 0}const auth_token=getCookie("auth",full_cookie);void 0===auth_token?console.log("Не получилось найти данные авторизации. Вы точно вошли в аккаунт?"):console.log(`\n\nCookie:\n\n${full_cookie}\n\nAuthorization:\n\n${parseAuthCookie(auth_token)}\n\n`);
```

При вставке кода в консоль, браузер может показать предупреждение о вставке кода. В таком случае, перед вставкой кода, введите в консоль то, что просит браузер, например: "разрешить вставку", нажмите Enter, а уже потом выполняйте вставку кода и снова нажимайте Enter.

В консоли появится текст вида:

```
Cookie:

<То, что нужно вставить в поле auth -> cookie конфига приложения>

Authorization:

<То, что нужно вставить в поле auth -> authorization конфига приложения>
```

Вставьте соответствующий текст в соответствующие поля конфига.

**Comment by JohnDoe-off** [comment]
Отлично! Всё работает. 
Пара вопросов автору.
1. Запоминает ли историю парсинга или каждый раз как первый раз?
2. Пишет ли что-то по окончанию парсинга?

**Reply by lowfc** [reply]
Добрый день. Если вы про JS скрипт из ответа выше - в браузере он ничего не сохраняет, просто получает куки и выводит в консоль. Но куки обновляются не так часто, поэтому указав их в конфиге приложения, вы какое-то время можете не получать их снова, пока авторизация не отлетит. 
По окончанию парсинга выводит либо:
```
Cookie:

<То, что нужно вставить в поле auth -> cookie конфига приложения>

Authorization:

<То, что нужно вставить в поле auth -> authorization конфига приложения>
```

Либо:
```
Не получилось найти данные авторизации. Вы точно вошли в аккаунт?
```

Не сжатую версию скрипта [можно посмотреть тут.](https://github.com/lowfc/boosty_downloader/blob/master/get_cookie_unminify.js)

**Reply by JohnDoe-off** [reply]
Я имел ввиду, что например если я закрыл программу и затем запустил парсинг повторно начнется ли он с места остановки, или если продолжить парсинг того же автора, то всё начнётся заново?
Хорошо бы конечно какую-то историю. А то объемы зачастую большие и на парсинг 300 гб уходит больше суток.

**Reply by lowfc** [reply]
В случае остановки синхронизации посередине процесса и повторном запуске, апишные запросы будут повторены в полном объеме (но они и не сильно тяжелые), а контент продолжит скачиваться с того места, на котором приложение остановилось в тот раз, т.к. скачивать по новой уже скачанные файлы приложение не станет. Однако, если приложение было закрыто в момент скачивания большого файла, и тот загрузиться не успел, то "дозагрузить" его приложение не сможет, и начнет скачивать сразу следующий файл. Лечится это удалением битого файла - тогда приложение загрузит его снова.

В теории конечно можно время от времени сохранять состояние текущей синхронизации, но, кажется, это не сильно ускорит процесс (и все сломается, если за время, пока синхронизация перезапускалась, на страницу автора загрузили что-то новое).

**Reply by JohnDoe-off** [reply]
Ясно. Спасибо.  
Было бы конечно здорово, если б в начале новой загрузки того же автора появилась красивая надпись.
- Continue?
- Upload again?
Потому что по большому счету, перебор скачанного сравним по времени с самой загрузкой.

**Reply by lowfc** [reply]
В целом можно добавить подобную оптимизацию, запишу в "бэклог".

**Reply by JohnDoe-off** [reply]
Спасибо!
